File: src/boj/boj_9012.java
package boj;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class boj_9012 {
    public static void main(String[] args) throws IOException {
        final BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        final int n = Integer.parseInt(br.readLine());
        final Stack<Character> stack = new Stack<>();
        for (int i = 0; i < n; i++) {
            stack.clear();
            final String input = br.readLine();
            int j;
            for (j = 0; j < input.length(); j++) {
                final char c = input.charAt(j);
                if (c =='(') {
                    stack.push(c);
                } else {
                    if (stack.isEmpty()) {
                        System.out.println("NO");
                        break;
                    }
                    stack.pop();
                }
            }
            if (j == input.length()) {
                if (stack.isEmpty()) {
                    System.out.println("YES");
                } else {
                    System.out.println("NO");
                }
            }
        }
    }
}

File: src/boj/boj_9012.kt
package boj

import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    val t = readLine().toInt()
    for (i in 0 until t) {
        var isFailed = false
        val stack = LinkedList<Char>()
        val string = readLine()
        for (c in string) {
            if (c == '(') {
                stack.add(c)
            } else {
                if (stack.isEmpty()) {
                    println("NO")
                    isFailed = true
                    break
                } else {
                    stack.poll()
                }
            }
        }
        if (isFailed) {
            continue
        } else if (stack.isNotEmpty()) {
            println("NO")
        } else {
            println("YES")
        }
    }
}